crates/uv-auth/src/middleware.rs

```diff
@@ -0,0 +1,185 @@
+            );
+            AuthenticationStore::set(&url, Some(auth));
+        } else if matches!(self.keyring_provider, KeyringProvider::Subprocess) {
+            // If we have keyring support enabled, we check there as well
+            match get_keyring_subprocess_auth(&url) {
+                Ok(Some(auth)) => {
+                    req.headers_mut().insert(
+                        reqwest::header::AUTHORIZATION,
+                        basic_auth(auth.username(), auth.password()),
+                    );
+                    AuthenticationStore::set(&url, Some(auth));
+                }
+                Ok(None) => {
+                    debug!("No keyring credentials found for {}", url);
+                }
+                Err(e) => {
+                    warn!("Failed to get keyring credentials for {}: {}", url, e);
+                }
+            }
+        }
+
+        // If we still don't have any credentials, we save the URL so we don't have to check netrc or keyring again
+        if !req.headers().contains_key(reqwest::header::AUTHORIZATION) {
+            AuthenticationStore::set(&url, None);
+        }
+
+        next.run(req, _extensions).await
+    }
+}
+
+/// Create a `HeaderValue` for basic authentication.
+///
+/// Source: <https://github.com/seanmonstar/reqwest/blob/2c11ef000b151c2eebeed2c18a7b81042220c6b0/src/util.rs#L3>
+fn basic_auth<U, P>(username: U, password: Option<P>) -> HeaderValue
+where
+    U: std::fmt::Display,
+    P: std::fmt::Display,
+{
+    use base64::prelude::BASE64_STANDARD;
+    use base64::write::EncoderWriter;
+    use std::io::Write;
+
+    let mut buf = b"Basic ".to_vec();
+    {
+        let mut encoder = EncoderWriter::new(&mut buf, &BASE64_STANDARD);
+        let _ = write!(encoder, "{}:", username);
+        if let Some(password) = password {
+            let _ = write!(encoder, "{}", password);
+        }
+    }
+    let mut header = HeaderValue::from_bytes(&buf).expect("base64 is always valid HeaderValue");
+    header.set_sensitive(true);
+    header
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use reqwest::Client;
+    use reqwest_middleware::ClientBuilder;
+    use std::io::Write;
+    use tempfile::NamedTempFile;
+    use wiremock::matchers::{basic_auth, method, path};
+    use wiremock::{Mock, MockServer, ResponseTemplate};
+
+    const NETRC: &str = r#"default login myuser password mypassword"#;
+
+    #[tokio::test]
+    async fn test_init() -> Result<(), std::io::Error> {
+        let server = MockServer::start().await;
+
+        Mock::given(method("GET"))
+            .and(path("/hello"))
+            .and(basic_auth("myuser", "mypassword"))
+            .respond_with(ResponseTemplate::new(200))
+            .mount(&server)
+            .await;
+
+        let status = ClientBuilder::new(Client::builder().build().unwrap())
+            .build()
+            .get(format!("{}/hello", &server.uri()))
+            .send()
+            .await
+            .unwrap()
```

Comment:
Should these unwraps just be `?` given that we already return `Result` here?

Reply:
Done